### PR TITLE
Retry mechanism for mirroring packages

### PR DESCRIPTION
**Pull request by jesusbv**
## Description

- Retry mirroring packages 5 times
  with 2 seconds delay between attempts

## Change Type

*Please select the correct option.*

- [ ] **Bug Fix** (a non-breaking change which fixes an issue)
- [X] **New Feature** (a non-breaking change which adds new functionality)
- [ ] **Documentation Update** (a change which only updates documentation)

## Checklist

*Please check off each item if the requirement is met.*

- [X] I have verified that my code follows RMT's coding standards with `rubocop`.
- [X] I have reviewed my own code and believe that it's ready for an external review.
- [ ] I have provided comments for any hard-to-understand code.
- [ ] I have documented the `MANUAL.md` file with any changes to the user experience.
- [X] RMT's test coverage remains at 100%.
- [X] If my changes are non-trivial, I have added a changelog entry to notify users at `package/obs/rmt-server.changes`.

## Other Notes

Please use this space to provide notes or thoughts to the team, such as tips on how to review/demo your changes.

### CODE REVIEW

**Comment by suse-tests-fail** [comment]
Prophet reports failure.
https://ci.suse.de/job/scc-RMT-integration-tests/252137/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment.
Failed fast in project rmt on step docker-compose-build-images. (Merged 7aa1e07a3c3e8d57c6c767967a2bc5074e537d4a into a5b45e86206dd3d56651550ab561c96264b8dc0f)

File: spec/lib/rmt/mirror_spec.rb [hunk 1]
```diff
@@ -800,6 +810,7 @@
           end
         end
 
+        # rmt_mirror.mirror(**mirror_params)
```
Comment (digitaltom) [inline spec/lib/rmt/mirror_spec.rb:813]: can we drop this line?

File: lib/rmt/mirror.rb [hunk 2]
```diff
@@ -85,7 +85,8 @@ def create_temp_dir
     raise RMT::Mirror::Exception.new(_('Could not create a temporary directory: %{error}') % { error: e.message })
   end
 
-  def mirror_metadata(repository_dir, repository_url, temp_metadata_dir, n_retry = MIRROR_METADATA_RETRIES)
+  def mirror_metadata(repository_dir, repository_url, temp_metadata_dir)
+    attempts ||= MIRROR_METADATA_RETRIES
```
Comment (digitaltom) [inline lib/rmt/mirror.rb:89]: I'd suggest to call this variable `remaining_attempts` to make clear that this is the remaining amount of retries. 
The constant `MIRROR_METADATA_RETRIES` could be called `MIRROR_RETRIES` because it's not only about metadata anymore.

File: lib/rmt/mirror.rb [hunk 3]
```diff
@@ -130,12 +123,19 @@ def mirror_metadata(repository_dir, repository_url, temp_metadata_dir, n_retry =
...
     metadata_files
   rescue StandardError => e
-    if n_retry == MIRROR_METADATA_RETRIES
-      raise RMT::Mirror::Exception.new(_('Error while mirroring metadata: %{error}') % { error: e.message })
-    end
+    message ||= e.message
+    raise RMT::Mirror::Exception.new(_('Error while mirroring metadata: %{error}') % { error: message }) if attempts == 0
+
+    attempts -= 1
+    n_seconds = 2
```
Comment (digitaltom) [inline lib/rmt/mirror.rb:130]: This could be extracted to a constant, then we don't need to define it 3 times.

File: lib/rmt/mirror.rb [hunk 4]
```diff
@@ -171,7 +179,15 @@ def mirror_packages(metadata_files, repository_dir, repository_url)
 
     raise _('Failed to download %{failed_count} files') % { failed_count: failed_downloads.size } unless failed_downloads.empty?
   rescue StandardError => e
-    raise RMT::Mirror::Exception.new(_('Error while mirroring data: %{error}') % { error: e.message })
+    message ||= e.message
```
Comment (digitaltom) [inline lib/rmt/mirror.rb:182]: Why is this line needed? Can the message get set from somewhere else?

File: lib/rmt/mirror.rb [hunk 5]
```diff
@@ -171,7 +179,15 @@ def mirror_packages(metadata_files, repository_dir, repository_url)
...
+    raise RMT::Mirror::Exception.new(_('Error while mirroring data: %{error}') % { error: message }) if attempts == 0
```
Comment (digitaltom) [inline lib/rmt/mirror.rb:183]: We could improve the message and write `Error while mirroring packages:` here?

**Comment by suse-tests-fail** [comment]
Prophet reports failure.
https://ci.suse.de/job/scc-RMT-integration-tests/252140/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment.
Failed fast in project rmt on step docker-compose-build-images.
Failed fast in project rmt on step docker-compose-build-images. (Merged 7aa1e07a3c3e8d57c6c767967a2bc5074e537d4a into b89b8d137cd61b8a7da0c01c7f5a73d5b3754833)

**Comment by suse-tests-fail** [comment]
Prophet reports failure.
https://ci.suse.de/job/scc-RMT-integration-tests/252154/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment.
Failed fast in project rmt on step docker-compose-build-images. (Merged 7aa1e07a3c3e8d57c6c767967a2bc5074e537d4a into b89b8d137cd61b8a7da0c01c7f5a73d5b3754833)

**Comment by suse-tests-fail** [comment]
Prophet reports failure.
https://ci.suse.de/job/scc-RMT-integration-tests/252680/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment.
Failed fast in project rmt on step docker-compose-build-images. (Merged 3ca29717c486f2d49894a77fe77fbedcb37b8b47 into b89b8d137cd61b8a7da0c01c7f5a73d5b3754833)

**Comment by suse-tests-fail** [comment]
Prophet reports failure.
https://ci.suse.de/job/scc-RMT-integration-tests/253498/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment.
Failed fast in project rmt on step docker-compose-build-images.
Failed fast in project rmt on step docker-compose-build-images.
Failed fast in project rmt on step docker-compose-build-images. (Merged 3ca29717c486f2d49894a77fe77fbedcb37b8b47 into b89b8d137cd61b8a7da0c01c7f5a73d5b3754833)

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/254530/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 3ca29717c486f2d49894a77fe77fbedcb37b8b47 into ce2325855246f8d9bf64a374ef5597a8c5b41dc7)

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/254558/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 3ca29717c486f2d49894a77fe77fbedcb37b8b47 into e1ba12c592c15f4f23401aaa1c9b31e62296cc2c)

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/254660/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 3ca29717c486f2d49894a77fe77fbedcb37b8b47 into c1ea57ba3c654075a14c59a440ee84ef6e0a548d)

File: lib/rmt/mirror.rb (hunk 4 above)
Comment (jesusbv) [reply lib/rmt/mirror.rb:182]: No, it can not. It is needed to get the error message that originated the exception

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/254812/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 8c903b62c29a77494a6cf52e1fae5f5e304dbd06 into c1ea57ba3c654075a14c59a440ee84ef6e0a548d)

File: lib/rmt/mirror.rb (hunk 4 above)
Comment (digitaltom) [reply lib/rmt/mirror.rb:181]: I was asking why this is `||=` instead of `=`. In which case would `message` already be set? Same in line 127.

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/254840/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 658d9f8e7bf292aa39f2b409b4f106ffcbb11b69 into c1ea57ba3c654075a14c59a440ee84ef6e0a548d)

File: lib/rmt/mirror.rb (hunk 4 above)
Comment (digitaltom) [reply lib/rmt/mirror.rb:181]: It turned out that the `||=` was hiding the current exceptions. In the tests this was hiding missing VCRs for example. 
I propose to go a different way and implement the retry login ion the `Downloader` class, where it feels better suited for me. I started to implement that here: https://github.com/SUSE/rmt/pull/907/

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/259508/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 658d9f8e7bf292aa39f2b409b4f106ffcbb11b69 into 454855918e87bc5de7d526e497708e2e57e75488)
